### PR TITLE
Generalize args to accept all and pass directly to samtools

### PR DESCRIPTION
**Pull request by jessebrennan**
This could be expanded to allow arbitrary args for the cram commands as well, but there are some complications due to the mismatched APIs of xsamtools and the underlying tools.

### CODE REVIEW

File: xsamtools/cli/cram.py [hunk 1]
```diff
@@ -6,7 +6,7 @@
 from xsamtools import cram
 
 
-def view(args: argparse.Namespace):
+def view(args: argparse.Namespace, extra_args):
```
Comment (DailyDreaming) [inline xsamtools/cli/cram.py:9]: Typing.

File: tests/test_xsamtools.py [hunk 2]
```diff
@@ -48,6 +48,25 @@ def test_combine(self):
                 output = "test_bcftools_combined.vcf.gz"
                 vcf.combine(inputs, output)
 
+    def test_cli_vcf_merge(self):
+        inputs = ["tests/fixtures/a.vcf.gz", "tests/fixtures/b.vcf.gz"]
+        output = "test_bcftools_combined.vcf.gz"
+        # Test with some arguments that get passed on to bcftools
+        vcf.combine(inputs, output, '--force-samples', '--print-header')
```
Comment (DailyDreaming) [inline tests/test_xsamtools.py:55]: Can this check that the expected output appeared and possibly check that the file is the expected size (and not, say, zero bytes)?

File: tests/test_xsamtools.py [hunk 3]
```diff
@@ -48,6 +48,25 @@ def test_combine(self):
...
+
+    def test_cli_subsample(self):
+        samples = ["NWD994242", "NWD637453"]
+        output_key = "test_bcftools_subsampled.vcf.gz"
+        src_path = "tests/fixtures/expected.vcf.gz"
+        dst_path = output_key
+        # Test with some arguments that get passed on to bcftools
+        vcf.subsample(src_path, dst_path, samples, '--no-header', '--no-version')
```
Comment (DailyDreaming) [inline tests/test_xsamtools.py:63]: Same as above.  Can this check that the expected output appeared and possibly check that the file is the expected size (and not, say, zero bytes)?

File: tests/test_xsamtools.py [hunk 4]
```diff
@@ -48,6 +48,25 @@ def test_combine(self):
...
+
+    def test_cli_stats(self):
+        src_path = "tests/fixtures/expected.vcf.gz"
+        # Test with some arguments that get passed on to bcftools
+        vcf.stats(src_path, '--1st-allele-only')
```
Comment (DailyDreaming) [inline tests/test_xsamtools.py:68]: Can this check the output of `vcf.stats`?

File: xsamtools/cli/vcf.py [hunk 5]
```diff
@@ -6,7 +6,7 @@
 from xsamtools import vcf
 
 
-def merge(args: argparse.Namespace):
+def merge(args: argparse.Namespace, extra_args):
```
Comment (DailyDreaming) [inline xsamtools/cli/vcf.py:9]: Typing.

File: xsamtools/cli/vcf.py [hunk 6]
```diff
@@ -16,23 +16,18 @@ def merge(args: argparse.Namespace):
...
-    extra_args = []
-    if args.force_samples:
-        extra_args.append('--force-samples')
-    if args.print_header:
-        extra_args.append('--print-header')
     vcf.combine(inputs, args.output, *extra_args)
 
 
-def subsample(args: argparse.Namespace):
+def subsample(args: argparse.Namespace, extra_args):
```
Comment (DailyDreaming) [inline xsamtools/cli/vcf.py:22]: Typing.

File: xsamtools/cli/vcf.py [hunk 7]
```diff
@@ -16,23 +16,18 @@ def merge(args: argparse.Namespace):
...
     """
     Subsample VCF a stored locally or in google bucket.
     """
-    vcf.subsample(args.input, args.output, args.samples.split(","))
+    vcf.subsample(args.input, args.output, args.samples.split(","), *extra_args)
 
 
-def stats(args: argparse.Namespace):
+def stats(args: argparse.Namespace, extra_args):
```
Comment (DailyDreaming) [inline xsamtools/cli/vcf.py:29]: Typing.

File: xsamtools/vcf.py [hunk 8]
```diff
@@ -35,13 +33,15 @@ def _view(input_filepath: str, output_filepath: str, samples: Sequence[str]):
                         "-O", "z",
                         "-S", tf.name,
                         "--threads", f"{2 * cores_available}",
+                        *args,
```
Comment (DailyDreaming) [inline xsamtools/vcf.py:36]: Should this check if `-O` or `-S` aren't already being supplied by the user and thus creating duplicate arguments?

File: xsamtools/vcf.py [hunk 9]
```diff
@@ -35,13 +33,15 @@ def _view(input_filepath: str, output_filepath: str, samples: Sequence[str]):
...
                         input_filepath],
                        check=True)
 
-def _stats(input_filepath: str):
+def _stats(input_filepath: str, *args):
     subprocess.run([samtools.paths['bcftools'],
                     "stats",
                     "--threads", f"{2 * cores_available}",
+                    *args,
```
Comment (DailyDreaming) [inline xsamtools/vcf.py:44]: Same as above, but for a potential duplicate `--threads` arg.

**Comment by DailyDreaming** [comment]
@jessebrennan It would also be useful to have negative tests that ensure that things break expectedly when a user supplies a nonsense argument that isn't recognized by either `bcftools` or `xsamtools`.

**Comment by jessebrennan** [comment]
@DailyDreaming great feedback! thanks I'll do so.

File: tests/test_xsamtools.py (hunk 4 above)
Comment (jessebrennan) [reply tests/test_xsamtools.py:68]: It has no output. Maye that's a bug, I don't really understand what the purpose of this tool is.

**Review by DailyDreaming — CHANGES_REQUESTED** [review]

File: xsamtools/cli/vcf.py (hunk 6 above)
Comment (DailyDreaming) [reply xsamtools/cli/vcf.py:22]: Still needs typing (another `: argparse.Namespace`)?

File: xsamtools/cli/vcf.py (hunk 5 above)
Comment (DailyDreaming) [reply xsamtools/cli/vcf.py:9]: Still needs typing (another `: argparse.Namespace`)?

File: xsamtools/cli/vcf.py (hunk 7 above)
Comment (DailyDreaming) [reply xsamtools/cli/vcf.py:29]: Still needs typing (another `: argparse.Namespace`)?

**Review by DailyDreaming — COMMENTED** [review]
Looks good, but I don't believe the typing comments were addressed.  Once those are addressed, this looks good to merge to me.

**Review by DailyDreaming — APPROVED** [review]
LGTM.  Thanks!

**Pull request merged**